### PR TITLE
Synch to mcu-tools/mcuboot of 6X2022

### PR DESCRIPTION
**Pull request by nvlsianpu**
Synchronized to:
https://github.com/mcu-tools/mcuboot/commit/4eca54f4175c76c4525ac0f1eb867ffb7d40e165

- add precise check of the image size
- loader: Added post copy hook to swap function
- Add Kconfig option for setting swap using move as default swap algorithm
- zephyr: fixed ram loading for ARM, with correct handling of vector table when code has moved to RAM.

- imgtool: add option to export public PEM

Merge using **Merge by Create a Merge commit!**

### CODE REVIEW

**Review by gmarull — APPROVED** [review]

File: boot/bootutil/src/loader.c [hunk 1]
```diff
@@ -1903,6 +1905,60 @@ boot_update_hw_rollback_protection(struct boot_loader_state *state)
...
+ *
+ * @param  state        Boot loader status information.
+ *
+ * @return              0 - image can be swapped, -1 downgrade prevention
+ */
+static int
+check_downgrade_prevention(struct boot_loader_state *state)
+{
+#if defined(MCUBOOT_DOWNGRADE_PREVENTION) && \
+    (defined(MCUBOOT_SWAP_USING_MOVE) || defined(MCUBOOT_SWAP_USING_SCRATCH))
```
Comment (nordicjm) [inline boot/bootutil/src/loader.c:1922]: In an upgrade only system e.g. nrf53 with netcore update support, assuming code is stored in external QSPI, doesn't that mean that if a person reloads an older firmware by flashing it to the QSPI directly then there is no way for mcuboot to block the overwrite? Or perhaps using serial recovery?

File: boot/zephyr/Kconfig [hunk 2]
```diff
@@ -537,14 +548,25 @@ choice BOOT_DOWNGRADE_PREVENTION_CHOICE
...
+config MCUBOOT_DOWNGRADE_PREVENTION_SECURITY_COUNTER
+	bool "Use image security counter instead of version number"
+	depends on MCUBOOT_DOWNGRADE_PREVENTION
+	depends on (BOOT_SWAP_USING_MOVE || BOOT_SWAP_USING_SCRATCH)
+	help
+       Security counter is used for version eligibility check instead of pure
+       version.  When this option is set, any upgrade must have greater or
+       equal security counter value.
+       Because of the acceptance of equal values it allows for software
+       downgrades to some extent.
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:568]: Wrongly indented, should be tab and 2 spaces

**Comment by nvlsianpu** [comment]
@nordicjm Objective for this PR is synchronization with given upstream  revision. It is not for reworking patches included.

File: boot/zephyr/Kconfig (hunk 2 above)
Comment (nvlsianpu) [reply boot/zephyr/Kconfig:568]: This might be fixed in mcu-tools/mcuboot, not here.

File: boot/bootutil/src/loader.c (hunk 1 above)
Comment (nvlsianpu) [reply boot/bootutil/src/loader.c:1922]: @nordicjm If someone has direct access to the HW at such level we can't do anything about it: this is out of mcuboot itself focus.
Serial recovery objective is giving the user total control on image(es).

**Review by de-nordic — APPROVED** [review]

**Pull request merged**